emukit/bayesian_optimization/acquisitions/expected_improvement.py

```diff
@@ -88,3 +90,338 @@ def get_standard_normal_pdf_cdf(x: np.array, mean: np.array, standard_deviation:
     return u, pdf, cdf
+
+
+class QExpectedImprovement(ExpectedImprovement):
+    def __init__(self, model: Union[IModel, IDifferentiable, IJointlyDifferentiable], jitter: float = float(0),
+                 fast_compute: bool = False, eps: float = 1e-3) -> None:
+        """
+        This acquisition computes for a given input the improvement over the current best observed value in
+        expectation for multiple points. For more information see:
+
+        Efficient Global Optimization of Expensive Black-Box Functions
+        Jones, Donald R. and Schonlau, Matthias and Welch, William J.
+        Journal of Global Optimization
+
+        Implementation based on paper and its source code:
+
+        Fast computation of the multipoint ExpectedImprovement with
+        applications in batch selection
+        Chevalier, C. and Ginsbourger, D.
+        International Conference on Learning and Intelligent Optimization.
+        Springer, Berlin, Heidelberg, 2013.
+
+        Source code: https://github.com/cran/DiceOptim
+
+        :param model: model that is used to compute the improvement.
+        :param jitter: parameter to encourage extra exploration.
+        :param fast_compute: Whether to use faster approximative method.
+        :param eps: Grid length for numerical derivative in approximative method.
+        """
+        super(QExpectedImprovement, self).__init__(model, jitter)
+        self.fast_compute = fast_compute
+        self.eps = eps
+
+    def evaluate(self, x: np.ndarray) -> np.ndarray:
+        """
+        Computes the multi point Expected Improvement.
+
+        :param x: points where the acquisition is evaluated.
+        """
+        mean, variance = self.model.predict(x, full_cov=True)
+        y_minimum = np.min(self.model.Y, axis=0)
+        return -QExpectedImprovement.get_acquisition(mean.flatten(), variance,
+                                                     y_minimum, self.fast_compute, self.eps)[0]
+
+    def evaluate_with_gradients(self, x: np.ndarray) -> Tuple:
+        """
+        Computes the Expected Improvement and its derivative.
+
+        :param x: locations where the evaluation with gradients is done.
+        """
+        mean, variance = self.model.predict(x, full_cov=True)
+        mean = mean.flatten()
+        y_minimum = np.min(self.model.Y, axis=0)
+        qei, pk, symmetric_term = QExpectedImprovement.get_acquisition(mean, variance, y_minimum,
+                                                                       self.fast_compute, self.eps)
+
+        mean_dx, variance_dx = self.model.get_joint_prediction_gradients(x)
+        qei_grad = QExpectedImprovement.get_acquisition_gradient(mean, variance, mean_dx, variance_dx, y_minimum,
+                                                                 pk, symmetric_term, self.fast_compute, self.eps)
+        return -qei, -qei_grad
+
+    @staticmethod
+    def get_acquisition(mu: np.ndarray, Sigma: np.ndarray, y_minimum: float,
```

Comment:
Would it be too much to ask to break this method down a bit? It's a bit too large tbh. Same goes for the next one

Reply:
Fixed now